python_ta/checkers/inconsistent_or_missing_returns_checker.py

```diff
@@ -0,0 +1,108 @@
+"""
+Check for inconsistent return statements in functions and missing return statements in non-None functions.
+"""
+
+from typing import Optional, Set
+
+from astroid import nodes
+from pylint.checkers import BaseChecker
+from pylint.checkers.utils import only_required_for_messages
+from pylint.lint import PyLinter
+
+
+class InconsistentReturnChecker(BaseChecker):
+
+    name = "inconsistent-or-missing-returns"
+    msgs = {
+        "R9710": (
+            """This function has inconsistent return statements: it sometimes returns a non-None value and sometimes uses `return` without a value.
+            To ensure clarity and consistency, every return statement should either return a value or explicitly return None""",
+            "inconsistent-returns",
+            "Used when a function has non-None return values at some cases and `return` in other cases",
+        ),
+        "R9711": (
+            "Missing return statement in function `%s`",
+            "missing-return-statement",
+            "Used when a function does not have a return statement and whose return type is not None",
+        ),
+    }
+
+    def __init__(self, linter: Optional[PyLinter] = None) -> None:
+        super().__init__(linter=linter)
+
+    @only_required_for_messages("missing-return-statement", "inconsistent-returns")
+    def visit_functiondef(self, node) -> None:
+        """Visit a function definition"""
+        self._check_return_statements(node)
+
+    def _check_return_statements(self, node) -> None:
+        """
+        Check for inconsistent returns and missing returns based on CFG
+        if the function has a return type annotation or has at least one
+        non-None return values.
+        """
+        if isinstance(node.returns, nodes.Const) and node.returns.value is None:
+            return
+
+        has_return_annotation = node.returns is not None
+        has_return_value = False
+
+        # get the blocks connected to the end of cfg
+        end = node.cfg.end
+        end_blocks = [edge.source for edge in end.predecessors]
+
+        # gather the return statement of each code block
+        return_statements = {}
+        for block in end_blocks:
+            return_statements[block] = None
+            statement = block.statements[-1]
+            if isinstance(statement, nodes.Return):
+                return_statements[block] = statement
+                if return_statements[block].value is not None:
+                    has_return_value = True
+
+        # check for inconsistent or missing returns
+        if has_return_annotation or has_return_value:
+            for block in return_statements:
+                statement = return_statements[block]
+                end_lines = self._search_for_end_line(block, set())
```

Comment:
There was a miscommunication here because I misspelled the attribute I wanted you to use.

PythonTA sets a custom attribute called `end_lineno` which should be correct for compound statements. Use that instead of `tolineno` (or `fromlineno`) to get the correct ending line for a statement without needing to do a graph traversal.

Reply:
Hi Professor, in fact using `end_lineno` attribute wouldn't work. The reason is that during the CFG construction, statements like `if` and loops are not included in `statements` in each block, as shown in the part of graph here.

<img width="290" alt="image" src="https://github.com/pyta-uoft/pyta/assets/147955278/3753cccf-eeb0-41e0-94be-f441767a98ef">

I will try to think of more efficient alternatives of graph traverse. One solution I can current think of is to turn on the `separate-condition-blocks` option of CFGVisitor, but I'm not sure whether it's safe to modify such a global parameter that could affect every CFG-based checker.

Reply:
@Raine-Yang-UofT okay, in this case I do not fully understand what the problem is. Please do the following:

1. Return to the previous approach you were using.
2. Write a test case that shows the incorrect behaviour.

After you push that I'll take a closer look.

Reply:
@david-yz-liu Hi Professor, the problem is more complicated than I though, as my implementation also works incorrectly at some cases like try/except statements. Before I proceed I want to first make sure where you want the comment `# INSERT RETURN STATEMENT` to appear.

For example, for this function, the outer `for` loop does not guarantee to have a return value
```python
def nested_statements(x):
    result = 0
    for i in range(x):
        result += i
        for j in range(0, 5):
            if j > 3:
                result += 1
            else:
                return result
```
Do we want the comment to be added below the `for` loop (which is how the code behaves if we pass `end_lineno` as the line number)

<img width="356" alt="image" src="https://github.com/pyta-uoft/pyta/assets/147955278/939f15d0-ce12-4764-bde2-44fd02f043a3">

Or, as I'm current trying to implement, adding the comment below the end of loop body

<img width="352" alt="image" src="https://github.com/pyta-uoft/pyta/assets/147955278/4a418940-b2d2-4761-adf2-7888091be78d">

Reply:
The expected behaviour is the latter. It's fine if you're a bit stuck on how to proceed. The reason I asked you to do the things in my last message is that this is the best way to provide enough information for me to help you.